File: runtime/java/units/rt/f32_max.java
// f32.max: a NaN operand yields wasm's canonical NaN (see f32_min).
// max(-0, +0) is +0.
static float f32_max(float a, float b) {
    if (Float.isNaN(a) || Float.isNaN(b)) {
        return Float.intBitsToFloat(0x7fc00000);
    }
    if (a > b) {
        return a;
    }
    if (b > a) {
        return b;
    }
    if (a == 0.0f) {
        // Equal zeros: max is -0 only if both operands are -0.
        return (Float.floatToRawIntBits(a) & Float.floatToRawIntBits(b)) < 0 ? -0.0f : 0.0f;
    }
    return a;
}

File: crates/dewasm-backend-go/units/rt/f32_max.go
func (rt) f32_max(a, b float32) float32 {
    // A NaN operand yields the wasm canonical NaN (see f32_min); Go's math.NaN() is not bit-canonical.
    if math.IsNaN(float64(a)) || math.IsNaN(float64(b)) {
        return math.Float32frombits(0x7fc00000)
    }
    if a > b {
        return a
    }
    if b > a {
        return b
    }
    if a == 0 {
        if !math.Signbit(float64(a)) || !math.Signbit(float64(b)) {
            return 0
        }
        return float32(math.Copysign(0, -1))
    }
    return a
}
